src/inverted_index.c

```diff
@@ -141,11 +141,123 @@ ENCODER(encodeDocIdsOnly) {
+
+  // The value is found as qint(docid,num).
+  // Signdedness is determined by the 3rd MSB in the header
+  NumEncoding_Float32 = 0x02 << 6,
+
+  // The value is found as qint32_64(docid, u64).
+  // If the number is signed, it is encoded as a double
+  NumEncoding_Extended = 0x03 << 6
+} NumEncodingType;
+
+#define NUM_TINY_MAX 0x3F            // Mask/Limit for 'Tiny' value
+#define NUM_F_SIGNED 0x20            // If encoding is not extended, signed bit
+#define NUM_ENCODING_MASK 0xC0       // Mask for two encoding bits
+#define NUM_QINT_VALUEHDR_MASK 0x3C  // Mask for qint value header
+
+typedef union {
+  float f32;
+  uint32_t encoded;
+} NumEncoded;
+
+static inline double myAbsf64(double input) {
+  return input < 0 ? -input : input;
+}
+
+static void dumpBits(uint64_t value, size_t numBits, FILE *fp) {
+  while (numBits) {
+    fprintf(fp, "%d", !!(value & (1 << (numBits - 1))));
+    numBits--;
+  }
+}
+
 // 9. Special encoder for numeric values
 ENCODER(encodeNumeric) {
-  size_t sz = WriteVarint(delta, bw);
+  const double absVal = myAbsf64(res->num.value);
+  const double realVal = res->num.value;
+
+  const uint32_t u32Num = (uint32_t)absVal;
+  const float f32Num = (float)absVal;
+  const uint64_t u64Num = (uint64_t)absVal;
+  const uint8_t u8Num = ((uint8_t)realVal) & NUM_TINY_MAX;
+  size_t sz = 0;
 
-  sz += Buffer_Write(bw, (char *)&res->num.encoded, sizeof(uint32_t));
+  if ((double)u8Num == realVal) {
+    uint8_t b = NumEncoding_Tiny | u8Num;
+    sz += Buffer_Write(bw, &b, 1);
+    sz += WriteVarint(delta, bw);
+    return sz;
+  }
+
+  uint8_t encoding = 0;
+  size_t pos = bw->buf->offset;
+
+  if ((double)u32Num == absVal) {
+    // -UINT32_MAX <= N <= UINT32_MAX
+    sz += qint_encode2(bw, delta, u32Num);
+    encoding = NumEncoding_Int32;
+    if (realVal < 0) {
+      encoding |= NUM_F_SIGNED;
+    }
+  } else if ((double)f32Num == absVal) {
+    // "Proper" floating point value. We can't really optimize this because
+    // of how floats are represented (i.e. exponent and mantissa), but we
+    // can optimize the sign bit.
+    NumEncoded num = {.f32 = f32Num};
+    if (realVal < 0) {
+      encoding |= NUM_F_SIGNED;
+    }
+
+    sz += qint_encode2(bw, delta, num.encoded);
+    encoding |= NumEncoding_Float32;
+
+  } else if ((double)u64Num == realVal) {
+    sz += qint_encode32_64pair(bw, delta, u64Num);
+    encoding |= NumEncoding_Extended;
+    // dumpBits(bw->buf->data[pos], 8, stdout);
+    // printf("\n");
+    // dumpBits(encoding, 8, stdout);
+    // printf("\n");
+  } else {
+    // All bets are off when trying to optimize encoding for a double that's
+    // actually a double and not an integral or floating point value.
+    // Can we do further optimizations here, though?
+    sz += Buffer_Write(bw, &encoding, 1);
+    sz += WriteVarint(delta, bw);
+    sz += Buffer_Write(bw, (void *)&realVal, 8);
+    encoding |= NumEncoding_Extended;
+  }
+
+  bw->buf->data[pos] |= encoding;
```

Comment:
This byte includes the qint header, which will always be <= 0b00111111. The encoding will only affect the upper two bits. If the encoding bits suggest that there are only two ints, we assume that only 0b00001111 is used for the qint header, and we use the 3rd MSB for the sign.